### PR TITLE
Components: Add ShortenedNumber component

### LINKED ISSUE #70853: Stats: Create a truncated number component
Image

Create a component for truncating numbers. This functionality is already used for the Annual Highlights, 7-day Highlights, and Best Views Ever components.

### PR DESCRIPTION
**Pull request by jsnmoon**
#### Proposed Changes

Refactors number shortening into a separate component.

#### Testing Instructions

**Storybook**
* Check out this branch locally and spin up Storybook via `yarn workspace @automattic/components run storybook`.
* Confirm that the "Highlight Cards" and "Shortened Number" stories look reasonable.

**Live Branch**
* Open the live branch for this PR and navigate to the Stats page.
* Ensure that the numbers look reasonable on the 7-day highlight section at the top of the Traffic page.
* Ensure that the numbers look reasonable on the annual and all-time highlight sections at the top of the Insights page.

#### Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

Fixes #70853.

### CODE REVIEW

File: client/my-sites/stats/all-time-highlights-section/index.tsx [hunk 1]
```diff
@@ -28,17 +32,9 @@ type MostPopularData = {
...
-	notation: 'compact',
-	compactDisplay: 'short',
-	maximumFractionDigits: 1,
-} );
-
-function formatNumber( number: number | null, isCompact = false ) {
-	const formatter = isCompact ? COMPACT_FORMATTER : FORMATTER;
-
-	return Number.isFinite( number ) ? formatter.format( number as number ) : '-';
+const FORMATTER = new Intl.NumberFormat();
```
Comment (grzegorz-cp) [inline client/my-sites/stats/all-time-highlights-section/index.tsx:35]: Shouldn't this keep a language parameter?

File: client/my-sites/stats/all-time-highlights-section/index.tsx [hunk 2]
```diff
@@ -148,7 +144,7 @@ export default function AllTimeHighlightsSection( { siteId }: { siteId: number }
 				{
 					id: 'views',
 					header: translate( 'Views' ),
-					content: formatNumber( viewsBestDayTotal, true ),
+					content: <TruncatedNumber value={ viewsBestDayTotal } />,
```
Comment (grzegorz-cp) [inline client/my-sites/stats/all-time-highlights-section/index.tsx]: Did you forget to replace the other places with `formatNumber()`, or was it intentional?

**Review by grzegorz-cp — COMMENTED** [review]
I've left a few questions.

File: client/my-sites/stats/all-time-highlights-section/index.tsx (hunk 1 above)
Comment (jsnmoon) [reply client/my-sites/stats/all-time-highlights-section/index.tsx:35]: I intentionally removed the first language parameter so [the formatter would default to the browser's locale](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/NumberFormat#syntax).

File: client/my-sites/stats/all-time-highlights-section/index.tsx (hunk 2 above)
Comment (jsnmoon) [reply client/my-sites/stats/all-time-highlights-section/index.tsx]: That was intentional; the remaining formatter in this file does not perform any shortening.

**Comment by jsnmoon** [comment]
Any ideas for a better component name? It just occurred to me that we're not performing any "truncation" with this component. We're merely shortening the number...

**Review by grzegorz-cp — APPROVED** [review]
Works as expected 👍

File: packages/components/src/shortened-number/index.tsx [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+	compactDisplay: 'short',
+	maximumFractionDigits: 1,
+} );
+
+function formatNumber( value: number | null ) {
+	if ( ! Number.isFinite( value ) ) {
+		return '-';
+	}
+
+	return Number.isFinite( value ) ? FORMATTER.format( value as number ) : '-';
```
Comment (dognose24) [inline packages/components/src/shortened-number/index.tsx:13]:
```suggestion
	return FORMATTER.format( value as number );
```

Could this line be simplified to return the formatted value only?

Comment (jsnmoon) [reply packages/components/src/shortened-number/index.tsx:13]: D'oh, good catch. Fixed.

**Pull request merged**